lib/pyfrc/physics/tankmodel.py

```diff
@@ -0,0 +1,316 @@
+                     nmotors, motor_config.name, gearing, wheel_diameter * 12.0)
+        
+        wheel_diameter *= FT_TO_M
+        robot_mass_kg = robot_mass * LB_TO_KG
+
+        max_velocity = ((motor_config.freeSpeedRpm / 60.0) * math.pi * wheel_diameter) / gearing
+        max_acceleration = (2.0 * nmotors * motor_config.stallTorque * gearing) / (wheel_diameter * robot_mass_kg)
+        
+        kv =  motor_config.nominalVoltage / max_velocity
+        ka = motor_config.nominalVoltage / max_acceleration
+        
+        logger.info("- Theoretical: vmax=%.3f, amax=%.3f, kv=%.3f, ka=%.3f",
+                    max_velocity*M_TO_FT, max_acceleration*M_TO_FT, kv, ka)
+        
+        return cls(motor_config,
+                   robot_mass, x_wheelbase,
+                   robot_width, robot_length,
+                   kv, ka, vintercept,
+                   kv, ka, vintercept,
+                   timestep_ms)
+    
+    def __init__(self, motor_config: MotorModelConfig,
+                       robot_mass: float,
+                       x_wheelbase: float,
+                       robot_width: float, robot_length: float,
+                       l_kv: float, l_ka: float, l_vi: float,
+                       r_kv: float, r_ka: float, r_vi: float,
+                       timestep_ms: int = 1):
+        '''
+            Use the constructor if you have measured ``kv``, ``ka``, and
+            ``Vintercept`` for your robot. Use the :func:`.theory` function
+            if you haven't.
+            
+            ``Vintercept`` is the minimum voltage required to generate enough
+            torque to overcome steady-state friction (see the paper for more
+            details).
+            
+            The robot width/length is used to compute the moment of inertia of
+            the robot. Don't forget about bumpers!
+        
+            :param motor_config: Motor specification
+            :param robot_mass:   Mass of robot (in lb)
+            :param x_wheelbase:  Wheelbase of the robot (in feet)
+            :param robot_width:  Width of the robot (in feet)
+            :param robot_length: Length of the robot (in feet)
+            :param l_kv:         Left side ``kv``
+            :param l_ka:         Left side ``ka``
+            :param l_vi:         Left side ``Vintercept``
+            :param r_kv:         Right side ``kv``
+            :param r_ka:         Right side ``ka``
+            :param r_vi:         Right side ``Vintercept``
+            :param timestep_ms:  Model computation timestep (in ms)
+        '''
+        
+        logger.info("Robot base: %.1fx%.1fin frame, %.1fin wheelbase, %.1flbs",
+                    robot_width * 12, robot_length * 12, x_wheelbase * 12, robot_mass)
+        
+        x_wheelbase *= FT_TO_M
+        robot_width *= FT_TO_M
+        robot_length *= FT_TO_M
+        
+        robot_mass *= LB_TO_KG
+        
+        self._lmotor = MotorModel(motor_config, l_kv, l_ka, l_vi)
```

Comment:
Ugh, so the ka/kv have units too. Maybe it's best if we just switch all of the units in this thing over to imperial units. At least it'll be consistent, which is probably more important.

Reply:
Well, given that _k<sub>v</sub>_ is voltage over velocity, and _k<sub>a</sub>_ is voltage over acceleration, shouldn't this have been obvious?

Anyway, I vote for SI units. If teams prefer to use imperial units, that should be explicit. There will be teams which will prefer to use SI units, and we shouldn't force them to use imperial units, especially given that the maths here works with SI units.

Reply:
It it were obvious to me, I would have originally considered that as a potential issue. 

I'm torn on this. While SI units are more 'right' -- given that a majority of the teams are in the US, and all of the field units are specified primarily in imperial units, I suspect most FRC high-school users would expect that and be more familiar with that.

Finally, [there's a CD thread for gathering data for this](https://www.chiefdelphi.com/forums/showthread.php?t=161539) and the units there are in US units. I suppose we could convert them.

It would be best to use a units library of some sort so that it didn't matter what users preferred... but in python I have to imagine that would be a performance hit.

Reply:
I think the solution for this (which you alluded to) is to get rid of units entirely, provide motor configurations in both imperial and SI units, and allow the users to use whatever makes sense to them. 

The question then is what units should stall torque be specified in? The SI unit is N-m, but on the CIM datasheet it's oz-in... which doesn't seem like it would be dimensionally correct. I assume the equivalent would be lb-ft, but I'm not able to reproduce the paper's computations for theoretical ka with that number.

Reply:
But from a dimensional point of view, that means that my other units (like robot weight) would need to be in ounces and (wheel diameter) would need to be in inches, right? Whereas if we used lb-ft, then you could stay in feet and pounds.

Reply:
Depends on how many conversions you want to make during your calculations. You can use any combination of these units.

Reply:
The goal would be to allow users to express things in units that feel natural without lots of conversions.

But, as I said, the numbers don't match the paper whether I use lbf or ozin or whatever, so that needs to be resolved. I posted a question on the paper's CD thread asking which number was used/intended, we'll see if the original author responds.